pkg/api/vlabs/types.go

```diff
@@ -555,12 +561,31 @@ func (m *MasterProfile) IsVirtualMachineScaleSets() bool {
 	return m.AvailabilityProfile == VirtualMachineScaleSets
 }
 
-// IsAllVirtualMachineScaleSets returns true if the cluster contains only Virtual Machine Scale Sets
-func (p *Properties) IsAllVirtualMachineScaleSets() bool {
-	isAll := p.MasterProfile.IsVirtualMachineScaleSets()
-	if isAll {
+// HasAvailabilityZones returns true if the master profile has availability zones
+func (m *MasterProfile) HasAvailabilityZones() bool {
+	return m.AvailabilityZones != nil && len(m.AvailabilityZones) > 0
+}
+
+// IsClusterAllAvailabilityZones returns true if the cluster contains AZs for all agents and masters profiles
+func (p *Properties) IsClusterAllAvailabilityZones() bool {
+	isAll := p.MasterProfile != nil && p.MasterProfile.HasAvailabilityZones()
```

Comment:
does this mean that this function should return false if MasterProfile is nil but all agents have AZs? Maybe it should be an || instead

What do you think about making this a single line? :
return  (p.MasterProfile == nil || p.MasterProfile.HasAvailabilityZones()) && p.HasZonesForAllAgentPools()

Reply:
this function should return true if the cluster contains all profiles with AZs. so in this case, if MasterProfile is nil then it should return false.

Reply:
> What do you think about making this a single line? :
return (p.MasterProfile == nil || p.MasterProfile.HasAvailabilityZones()) && p.HasZonesForAllAgentPools()

how about 
```
return (p.MasterProfile != nil && p.MasterProfile.HasAvailabilityZones()) && p.HasZonesForAllAgentPools())
```

Reply:
So that means it won't be possible to use AZ for agent pool only (AKS) clusters? The way I see this if a cluster only has agents and all its agents use AZ then all the cluster nodes use AZ. If that's not the logic we are trying to achieve let's rename the function to something like `MastersAndAgentsUseAvailabilityZones`

Reply:
> AZ for agent pool only (AKS) clusters

That is the next thing we want to focus on after we have enabled AZ for masters and agents as it is done currently. 

> The way I see this if a cluster only has agents and all its agents use AZ then all the cluster nodes use AZ

This PR currently has validations to ensure all profiles (masters and agents) use AZ in the cluster. We do not allow a mix of AZ and non-AZ when creating a new cluster. Zones need to be set explicitly configured hence the requirement to ask the user to configure AZ for both masters and agents. Let me know if you have other questions or recommendations. 

----

To avoid confusion, I have added the following for the `availabilityZones` feature as documented in `clusterdefinition.md`. Please let me know if any of this is not clear or can be improved.

To protect your cluster from datacenter-level failures, you can enable the Availability Zones feature for your cluster by configuring `"availabilityZones"` for the master profile and all of the agentPool profiles in the cluster definition. This feature only applies to Kubernetes clusters version 1.12+. Supported values are arrays of strings, each representing a supported availability zone in a region for your subscription. e.g. `"availabilityZones": ["1","2"]` represents zone 1 and zone 2 can be used. To get supported zones for a region in your subscription, run `az vm list-skus --location centralus --query "[?name=='Standard_DS2_v2'].[locationInfo, restrictions"] -o table`. You should see values like `'zones': ['2', '3', '1']` appear in the first column. If `NotAvailableForSubscription` appears in the output, then create an Azure support ticket to enable zones for that region. NOTE: To ensure high availability, each profile must define at least two nodes per zone. e.g. An agent pool profile with 2 zones: `"availabilityZones": ["1","2"]` must have at least 4 nodes total with `"count": 4`. When `"availabilityZones"` is configured, the `"loadBalancerSku"` will default to `Standard` as Standard LoadBalancer is required for availability zones.